packages/server/src/app/db/notifications/notifications.service.ts

```diff
@@ -0,0 +1,415 @@
+                const subscribers = await this.subscriptionModel.find(
+                    { notificationSourceId: toPublish.sourceId, notificationKind: toPublish.kind }
+                );
+                
+                // Publish this notification to all subscribers.
+                const publishableNotifications: NotificationDocument[] = subscribers.map(x =>
+                    this.getSpecificNotificationModel(x, toPublish)
+                );
+                await this.notificationModel.insertMany(publishableNotifications);
+
+                // Mark the notification in the NotificationQueue collection as "Published"
+                toPublish.publishStatus = PublishStatus.Published;
+                await toPublish.save();
+
+                processed += 1;
+                if (processed >= numToProcess) {
+                    break;
+                }
+
+                // If we've exceeded our deadline, end early and sleep
+                const [sec, ns] = process.hrtime(tickStart);
+                const elapsedMs = (sec * NS_PER_SEC + ns) / NS_PER_MS;
+                if (elapsedMs > MAX_MS_PER_WAKEUP) {
+                    break;
+                }
+            }
+            catch (error) {
+                if (error instanceof Error) {
+                    console.error(`Error processing notification from queue - ${error.name}: ${error.message}\n${error.stack}`);
+                } else {
+                    console.error(`Error processing notification from queue - ${error}`);
+                }
+                
+                break;
+            }
+        } // end while-loop
+
+        // Put stale notifications back into the processing queue
+        await this.cleanUpStaleNotifications();
+
+        // Queue the next wakeup        
+        setTimeout(
+            async () => {                
+                this.currentWakeup = this.processUnpublishedNotifications(MAX_NOTIFICATIONS_PER_WAKEUP);
+                await this.currentWakeup;
+            },
+            5000);
+    }
+
+    /** Put unpublished notifications that have been "In Progress" for more than 10 seconds back into the queue. */
+    private async cleanUpStaleNotifications(): Promise<void> {
+        const nowMinus10s = new Date(Date.now() - (10 * 1000));
+        const staleNotifications: NotificationQueueDocument[] = await this.notificationQueueModel.find(
+            { publishStatus: PublishStatus.InProgress, updatedAt: { $lte: nowMinus10s } }
+        );
+        await Promise.all(staleNotifications.map(x => {
+            x.publishStatus = PublishStatus.NotStarted;
+            return x.save();
+        }));
+    }
+
+    private getSpecificNotificationModel(subscription: NotificationSubscriptionDocument, toPublish: NotificationQueueDocument): NotificationDocument {
+        const commonProperties: NotificationBase = {
```

Comment:
This method is the last bit of really gnarly boilerplate. I tried a few different ways to cut it down to something a little more OOPy, or less horrible, to no avail. This was the least-worst I could get it.